package mechaverse;

option java_package = "org.mechaverse.api.proto.simulation.ant";
option java_outer_classname = "AntSimulation";

message EntityStats {
  // Age of the entity in environment update cycles.
  optional int64 age = 1;
  optional int64 energy_expended = 2;
  optional int64 distance_traveled = 3;
}

enum Direction {
  EAST = 0;
  NORTH_EAST = 1;
  NORTH = 2;
  NORTH_WEST = 3;
  WEST = 4;
  SOUTH_WEST = 5;
  SOUTH = 6;
  SOUTH_EAST = 7;
}
  
message EntityState {
  // Unique UUID for the entity.
  required string id = 1;
  required int32 x = 2;
  required int32 y = 3;
  
  // Direction the entity is facing in degrees.
  optional Direction direction = 4;
  // Current number of energy units.
  optional int32 energy = 5;
  // Maximum number of energy units.
  optional int32 max_energy = 6;
  
  optional EntityStats stats = 100;
}

message Ant {
  
  message InputPositions {
    optional uint32 random = 1;
    optional uint32 general = 2;
    optional uint32 energy_level = 3;
    optional uint32 current_cell_entity_sensor = 4;
    optional uint32 current_cell_entity_id_sensor = 5;
    optional uint32 front_entity_sensor = 6;
    optional uint32 front_entity_id_sensor = 7;
    optional uint32 front_left_sensor = 8;
    optional uint32 front_right_sensor = 9;
    optional uint32 left_sensor = 10;
    optional uint32 right_sensor = 11;
    optional uint32 pheromone_sensor = 12;
    optional uint32 direction_sensor = 13;
  }
  
  message OutputPositions {
    optional uint32 general = 1;
    optional uint32 move = 2;
    optional uint32 turn = 3;
    optional uint32 pick_up = 4;
    optional uint32 drop = 5;
    optional uint32 leave_pheromone = 6;
    optional uint32 interact = 7;
    optional uint32 trophallaxis = 8;
  }

  required EntityState entity_state = 1;
  optional string carried_entity_id = 2;
  
  optional bytes data = 3;
  optional InputPositions input_positions = 4;
  optional OutputPositions output_positions = 5;
   
  extensions 100 to max;
}

message Barrier {
  required EntityState entity_state = 1;
}

message Conduit { 
  required EntityState entity_state = 1;
  required string target_environment_id = 2;
}

message Dirt {
  required EntityState entity_state = 1;
}

message Food {
  required EntityState entity_state = 1;
}

message Pheromone {
  required EntityState entity_state = 1;
  optional int32 value = 2;
}

message Rock {
  required EntityState entity_state = 1;
}

message Environment {

  message EnvironmentStats {
    // Age of the environment in update cycles.
    optional int64 age = 1;
  }

  // Unique UUID for the environment.
  required string id = 1;
  required int32 width = 2;
  required int32 height = 3;
  optional EnvironmentStats stats = 4;
  
  // Entities.
  repeated Ant ant = 100;
  repeated Barrier barrier = 101;
  repeated Conduit conduit = 102;
  repeated Dirt dirt = 103;
  repeated Food food = 104;
  repeated Pheromone pheromone = 105;
  repeated Rock rock = 106;
}

message SimulationState {
  required string id = 1;
  required Environment environment = 2;
  repeated Environment sub_environment = 3;
}